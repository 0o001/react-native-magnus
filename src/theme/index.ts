export { ThemeContext } from './themeContext';
export { withTheme } from './withTheme';
export { ThemeProvider, ThemeProviderProps } from './themeProvider.component';

export {
  ThemeType,
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from './type';
