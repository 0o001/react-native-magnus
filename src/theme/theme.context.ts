import * as React from 'react';

import { ThemeType } from './type';

const defaultValue: ThemeType = {};

export interface ThemeContextType {
  theme: ThemeType;
  setTheme?: (theme?: ThemeType) => void;
}

export const THEME_CONTEXT_DEFAULT_VALUE = {
  theme: defaultValue,
};

export const ThemeContext: React.Context<ThemeContextType> = React.createContext(
  THEME_CONTEXT_DEFAULT_VALUE
);
