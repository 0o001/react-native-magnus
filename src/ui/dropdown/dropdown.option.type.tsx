import { ButtonProps } from '../button/button.type';

export interface OptionProps extends ButtonProps {
  value: any;
  onSelect?: (value: any) => void;
}
