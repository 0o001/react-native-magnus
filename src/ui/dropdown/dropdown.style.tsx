import { StyleSheet } from 'react-native';

import {
  getThemeProperty,
  createSpacingStyles,
  createBorderWidthStyles,
  createBorderColorStyles,
  createBorderRadiusStyles,
} from '../../theme/theme.service';

/**
 * computed style
 *
 * @param theme
 * @param props
 */
export const getStyle = (theme: any, props: any) => {
  const computedStyle: any = {};

  computedStyle.wrapper = {
    backgroundColor: getThemeProperty(theme.colors, props.bg, 'transparent'),
    ...createBorderWidthStyles(props),
    ...createBorderColorStyles(props, theme.colors),
    ...createBorderRadiusStyles(props, theme.borderRadius),
  };

  computedStyle.indicator = {
    alignSelf: 'center',
    marginVertical: 10,
  };

  computedStyle.container = {
    ...createSpacingStyles(props, theme.spacing),
  };

  if (props.h) {
    computedStyle.container = {
      ...computedStyle.container,
      height: props.h,
    };
  }

  // merging style props to computed style
  if (props.style) {
    computedStyle.container = {
      ...computedStyle.container,
      ...props.style,
    };
  }

  return StyleSheet.create(computedStyle);
};
