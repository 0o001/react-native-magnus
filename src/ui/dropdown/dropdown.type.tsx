import { ModalProps } from '../modal/modal.type';

import {
  BorderPropsType,
  SpacingPropsType,
  RoundedPropsType,
} from '../../theme';

export interface DropdownRef {
  open: () => void;
  close: () => void;
}

export interface DropdownProps
  extends ModalProps,
    BorderPropsType,
    SpacingPropsType,
    RoundedPropsType {
  title?: string | React.ReactNode;
  bg?: string;
  h?: number | string;
  w?: number | string;
  minW?: number | string;
  minH?: number | string;
  showSwipeIndicator?: boolean;
  justifyContent?:
    | 'flex-start'
    | 'flex-end'
    | 'center'
    | 'space-between'
    | 'space-around'
    | 'space-evenly';
  overflow?: 'hidden' | 'visible' | 'scroll';
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'stretch' | 'baseline';
  flexDir?: 'row' | 'column' | 'row-reverse' | 'column-reverse';
  flexWrap?: 'wrap' | 'nowrap' | 'wrap-reverse';
  children: React.ReactElement[] | React.ReactElement;
}
