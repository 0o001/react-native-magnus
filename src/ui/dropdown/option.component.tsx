import * as React from 'react';

import { GestureResponderEvent as RNGestureResponderEvent } from 'react-native';
import { OptionProps } from './option.type';
import { Button } from '../button/button.component';

const Option: React.FunctionComponent<OptionProps> = props => {
  const { children, value, onPress: onPressProp, onSelect, ...rest } = props;

  /**
   * on press select option
   *
   * @param e
   */
  const onPress = (event: RNGestureResponderEvent) => {
    if (onSelect) {
      onSelect(value);
    }

    if (onPressProp) {
      onPressProp(event);
    }
  };

  return (
    <Button {...rest} onPress={onPress} block>
      {children}
    </Button>
  );
};

Option.defaultProps = {
  onSelect: () => {},
  rounded: 0,
  bg: 'white',
  p: 0,
  color: 'black',
  alignItems: 'flex-start',
  justifyContent: 'flex-start',
  center: false,
};

export { Option };
