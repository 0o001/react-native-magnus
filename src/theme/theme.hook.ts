import React, { useState } from "react";

import { ThemeType } from "./type";
import { computeTheme } from "./theme.service";
import { ThemeContextType } from "./theme.context";

export const useTheme = (initialTheme?: ThemeType): ThemeContextType => {
  const [theme, setTheme] = useState(computeTheme(initialTheme));

  const setCurrentTheme = React.useCallback((newTheme?: ThemeType): void => {
    setTheme(computeTheme(newTheme));
  }, []);

  return {
    theme,
    setTheme: setCurrentTheme
  };
};
