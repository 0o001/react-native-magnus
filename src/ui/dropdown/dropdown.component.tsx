import * as React from 'react';
import { useContext, useState, useImperativeHandle } from 'react';
import { SafeAreaView } from 'react-native';
import RNModal from 'react-native-modal';

import { Div } from '../div/div.component';
import { Text } from '../text/text.component';
import { ThemeContext, getThemeProperty } from '../../theme';
import { getStyle } from './dropdown.style';
import { DropdownProps, DropdownRef } from './dropdown.type';

const Dropdown = React.forwardRef<DropdownRef, DropdownProps>((props, ref) => {
  const {
    bg,
    m,
    mt,
    mr,
    mb,
    ml,
    p,
    pr,
    pt,
    pb,
    pl,
    rounded,
    roundedTop,
    roundedRight,
    roundedBottom,
    roundedLeft,
    borderColor,
    borderBottomColor,
    borderLeftColor,
    borderTopColor,
    borderRightColor,
    borderWidth,
    borderLeftWidth,
    borderRightWidth,
    borderBottomWidth,
    borderTopWidth,
    title,
    showSwipeIndicator,
    children,
    backdropColor,
    backdropOpacity,
    flexDir,
    flexWrap,
    ...rest
  } = props;
  const { theme } = useContext(ThemeContext);
  const computedStyle = getStyle(theme, props);
  const [isVisible, setIsVisible] = useState(false);

  /**
   * exposing functions to parent
   */
  useImperativeHandle(ref, () => ({
    open() {
      setIsVisible(true);
    },
    close() {
      setIsVisible(false);
    },
  }));

  /**
   * render title
   */
  const renderTitle = () => {
    if (title) {
      return typeof title === 'string' ? (
        <Text fontSize="text400" color="gray700">
          {title}
        </Text>
      ) : (
        title
      );
    }

    return false;
  };

  /**
   * render scroll indicator
   */
  const renderIndicator = () => {
    return (
      showSwipeIndicator && (
        <Div
          h={7}
          rounded="xl"
          w={40}
          bg="gray200"
          style={computedStyle.indicator}
        />
      )
    );
  };

  return (
    <RNModal
      backdropTransitionOutTiming={0}
      isVisible={isVisible}
      onSwipeComplete={() => setIsVisible(false)}
      swipeDirection={['down']}
      backdropOpacity={backdropOpacity}
      backdropColor={getThemeProperty(
        theme.colors,
        backdropColor,
        'transparent'
      )}
      onBackdropPress={() => setIsVisible(false)}
      style={{
        margin: 0,
        justifyContent: 'flex-end',
      }}
      {...rest}
    >
      <Div style={computedStyle.wrapper}>
        {renderIndicator()}
        <SafeAreaView>
          <Div style={computedStyle.container}>
            {renderTitle()}
            <Div style={computedStyle.options}>
              {React.Children.map(children, (child: React.ReactElement) => {
                return React.cloneElement(child, {
                  onSelect: () => {
                    setIsVisible(false);
                  },
                });
              })}
            </Div>
          </Div>
        </SafeAreaView>
      </Div>
    </RNModal>
  );
});

Dropdown.defaultProps = {
  bg: 'white',
  rounded: 'none',
  showSwipeIndicator: true,
  backdropColor: 'gray900',
  backdropOpacity: 0.5,
  flexWrap: 'nowrap',
};

export { Dropdown };
