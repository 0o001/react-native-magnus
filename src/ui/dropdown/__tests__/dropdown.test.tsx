import React from "react";
import renderer from "react-test-renderer";

import { Dropdown } from "../dropdown.component";
import { Option } from "../option.component";
import { ThemeProvider } from "../../../theme";

describe("Dropdown Component", () => {
  it("should render without issues", () => {
    const instance = renderer.create(
      <ThemeProvider>
        <Dropdown onSelect={() => {}}>
          <Option value={1} />
        </Dropdown>
      </ThemeProvider>
    );

    expect(instance.toJSON()).toMatchSnapshot();
  });
});
