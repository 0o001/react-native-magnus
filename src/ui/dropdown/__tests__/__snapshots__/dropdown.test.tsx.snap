// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown Component should render without issues 1`] = `
<Modal
  animationType="none"
  hardwareAccelerated={false}
  hideModalContentWhileAnimating={false}
  onModalHide={[Function]}
  onModalWillHide={[Function]}
  onModalWillShow={[Function]}
  onRequestClose={[Function]}
  onSwipeComplete={[Function]}
  scrollHorizontal={false}
  scrollOffset={0}
  scrollOffsetMax={0}
  scrollTo={null}
  supportedOrientations={
    Array [
      "portrait",
      "landscape",
    ]
  }
  swipeDirection={
    Array [
      "down",
    ]
  }
  swipeThreshold={100}
  transparent={true}
  visible={false}
>
  <View
    accessible={true}
    clickable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 750,
      }
    }
  />
  <View
    hideModalContentWhileAnimating={false}
    onModalHide={[Function]}
    onModalWillHide={[Function]}
    onModalWillShow={[Function]}
    onMoveShouldSetResponder={[Function]}
    onMoveShouldSetResponderCapture={[Function]}
    onResponderEnd={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderReject={[Function]}
    onResponderRelease={[Function]}
    onResponderStart={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    onStartShouldSetResponderCapture={[Function]}
    onSwipeComplete={[Function]}
    pointerEvents="box-none"
    scrollHorizontal={false}
    scrollOffset={0}
    scrollOffsetMax={0}
    scrollTo={null}
    style={
      Object {
        "flex": 1,
        "justifyContent": "flex-end",
        "left": 0,
        "margin": 0,
        "top": 0,
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
      }
    }
    supportedOrientations={
      Array [
        "portrait",
        "landscape",
      ]
    }
    swipeDirection={
      Array [
        "down",
      ]
    }
    swipeThreshold={100}
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 0,
          "flexDirection": "column",
          "flexWrap": "nowrap",
          "justifyContent": undefined,
          "position": "relative",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-start",
            "alignSelf": "center",
            "backgroundColor": "#d9d9d9",
            "borderRadius": 12,
            "flexDirection": "column",
            "flexWrap": "nowrap",
            "height": 7,
            "justifyContent": undefined,
            "marginVertical": 10,
            "position": "relative",
            "width": 40,
          }
        }
      />
      <RCTSafeAreaView
        emulateUnlessSupported={true}
        pointerEvents="box-none"
      >
        <View
          style={
            Object {
              "alignItems": "flex-start",
              "backgroundColor": "#FFFFFF",
              "borderRadius": 0,
              "flexDirection": "column",
              "flexWrap": "nowrap",
              "justifyContent": undefined,
              "position": "relative",
            }
          }
        >
          <View
            accessible={true}
            clickable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "alignSelf": "stretch",
                "backgroundColor": "#FFFFFF",
                "borderRadius": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "overflow": "hidden",
                "padding": 0,
                "position": "relative",
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "alignSelf": "stretch",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                  "position": "relative",
                  "width": "100%",
                }
              }
            />
          </View>
        </View>
      </RCTSafeAreaView>
    </View>
  </View>
</Modal>
`;
